src/crypto/equihash.cpp

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) 2016 Jack Grigg
+// Copyright (c) 2016 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "crypto/equihash.h"
+#include "util.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <stdexcept>
+
+void validate_params(int n, int k)
+{
+    if (k>=n) {
+        std::cerr << "n must be larger than k\n";
+        throw invalid_params();
+    }
+    if (n % 8 != 0) {
+        std::cerr << "Parameters must satisfy n = 0 mod 8\n";
+        throw invalid_params();
+    }
+    if ((n/(k+1)) % 8 != 0) {
+        std::cerr << "Parameters must satisfy n/(k+1) = 0 mod 8\n";
+        throw invalid_params();
+    }
+}
+
+StepRow::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i) :
+        hash {new unsigned char[n/8]},
+        len {n/8},
+        indices {i}
+{
+    eh_HashState state;
+    state = base_state;
+    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
+    crypto_generichash_blake2b_final(&state, hash, n/8);
```

Comment:
The hash state type is abstracted but these calls are not; I suggest having a #define macro or a constant to refer to crypto_generichash_blake2b_init/update/final, e.g.:
```
#define eh_hash(x) crypto_generichash_blake2b_##x
```